### PR TITLE
Filebeat new registry

### PR DESCRIPTION
**Pull request by urso**
This PR updates the current filebeat registry to use libbeat/registry as backend. This is the same new registry to be used with #14229. This way inputs using the old input API and inputs using the new inputs API will use the very same registry file.

The reason for having both using the same backend is to ease migration. This PR introduces some initial support for registry file migration (in case we update the backend in the future). The migration support already copies the old registry file into the new location. By migrating the old file from the get go we do not have to deal with 2 different file formats now and in the future.

Mixing the registrar with the new registry is not exactly optimal. Traditionally filebeat stores states per input and in the registrar. Now we need to store it in the input, the registrar, and the new registry (also in-memory). Plus the new registry supports selective field updates, but the registrar doesn't make use of it. This means all fields need to be serialized and written twice. Once for generic in memory state and once for writing to disk. Although in most cases only one field would be enough, we need to serialize all fields all the time, due to the registrar mixing internal and external state in one common structure. This can't be really helped and is a temporary limitation.

To get an idea about the impact I did run a few tests. Each test runs filebeat for ~90s and we measure throughput, memory, and CPU usage (besides small peaks and lows run was rather stable). NASA HTTP access logs are chosen as sample data.

1. Filebeat -> console. Single log file.

```       Throughput    CPU    RSS  
master       76k/s     233%   50.5MB
newreg       73k/s     277%   49MB
```

2. Filebeat -> console, 101 log files.

```       Throughput    CPU    RSS  
master       83.6k/s   243%   46.5MB
newreg       83k/s     279%   46.2MB
```

3. Filebeat -> console, 101 log files, registry flush timeout 2s

```       Throughput    CPU    RSS  
master       92.5k/s   258%   44.8MB
newreg       92k/s     261%   42.5MB
```

All in all there is a performance degragation due to this change. This becomes most noticable in increased CPU usage. Each update serializes the complete state, no matter if only one field is updated. With separating meta-data and read-offset updates in the future I'd expect the overhead for serializing and writing updates to be reduced.

### CODE REVIEW

**Comment by ph** [comment]
@urso is that a draft? I've put the work in progress label on it.

File: libbeat/registry/registry.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+type Registry struct {
+	backend backend.Registry
+
+	mu     sync.Mutex
+	active map[string]*sharedStore // active/open stores
+	wg     sync.WaitGroup
+}
+
+// Key is the type used for all keys in a store.
+// type Key []byte
```
Comment (kvch) [inline libbeat/registry/registry.go]: This line seems unnecessary.

File: libbeat/registry/registry.go [hunk 2]
```diff
@@ -46,7 +46,7 @@ type ValueDecoder interface {
 }
 
 // NewRegistry creates a new Registry with a configured backend.
```
Comment (kvch) [inline libbeat/registry/registry.go:48]:
```suggestion
// New creates a new Registry with a configured backend.
```

File: filebeat/registrar/migrate.go [hunk 3]
```diff
@@ -26,15 +26,51 @@ import (
...
+		permissions: cfg.Permissions,
+	}
+}
+
+func (m *Migrator) Run() error {
+	return ensureCurrent(m.dataPath, m.migrateFile, m.permissions)
+}
+
+// ensureCurrent migrates old registry versions to the most recent version.
 func ensureCurrent(home, migrateFile string, perm os.FileMode) error {
```
Comment (kvch) [inline filebeat/registrar/migrate.go:74]: Why isn't this function a private function of `Migrator`?
```go
func (m *Migrator) ensureCurrent() error {
```
And looking at the other functions in this file, those can be part of the `Migrator` as well.

File: filebeat/registrar/migrate.go [hunk 4]
```diff
@@ -119,31 +171,110 @@ func initRegistry(regHome string, perm os.FileMode) error {
...
 }
 
-func readVersion(regHome, migrateFile string) (string, error) {
+// migrateVersion1 migrates the filebeat registry from version 0 to version 1
+// only. Version 1 is based on the implementation of version 1 in
+// libbeat/registry/backend/memlog.
+func migrateVersion1(home, regHome string, perm os.FileMode) error {
+	logp.Info("Migrate registry version 0 to version 1")
+
+	origDataFile := filepath.Join(regHome, "data.json")
```
Comment (kvch) [inline filebeat/registrar/migrate.go:180]: Nit: Could you please rename `origDataFile` to `originalDataFile`. It is a bit confusing to me. I was not sure whether it is supposed to be `original` or `origin`.

**Pull request closed without merging**